examples/D3D12Binding.cpp

```diff
@@ -152,14 +172,47 @@ class D3D12Binding : public BackendBinding {
         ComPtr<IDXGISwapChain3> swapChain;
         ComPtr<ID3D12DescriptorHeap> renderTargetViewHeap;
         ComPtr<ID3D12Resource> renderTargetResources[kFrameCount];
-        uint64_t fenceValues[kFrameCount] = { 0, 0 };
+        ComPtr<ID3D12CommandAllocator> commandAllocators[kFrameCount];
+        ComPtr<ID3D12GraphicsCommandList> commandList;
+        uint64_t fenceValues[kFrameCount] = {0};
+        uint64_t currentFrameNumber;
         ComPtr<ID3D12Fence> fence;
         HANDLE fenceEvent;
 
-        void SetNextRenderTarget() {
+        void SetNextRenderTarget(uint32_t frameIndex) {
             D3D12_CPU_DESCRIPTOR_HANDLE renderTargetViewHandle = renderTargetViewHeap->GetCPUDescriptorHandleForHeapStart();
             renderTargetViewHandle.ptr += rtvDescriptorSize * frameIndex;
-            backend::d3d12::SetNextRenderTarget(backendDevice, renderTargetResources[frameIndex], renderTargetViewHandle);
+            backend::d3d12::SetNextRenderTargetDescriptor(backendDevice, renderTargetViewHandle);
+
+            D3D12_RESOURCE_BARRIER resourceBarrier;
+            resourceBarrier.Transition.pResource = renderTargetResources[frameIndex].Get();
+            resourceBarrier.Transition.StateBefore = D3D12_RESOURCE_STATE_PRESENT;
+            resourceBarrier.Transition.StateAfter = D3D12_RESOURCE_STATE_RENDER_TARGET;
+            resourceBarrier.Transition.Subresource = D3D12_RESOURCE_BARRIER_ALL_SUBRESOURCES;
+            resourceBarrier.Type = D3D12_RESOURCE_BARRIER_TYPE_TRANSITION;
+            resourceBarrier.Flags = D3D12_RESOURCE_BARRIER_FLAG_NONE;
+            commandList->ResourceBarrier(1, &resourceBarrier);
+
+            ASSERT_SUCCESS(commandList->Close());
+            ID3D12CommandList* commandLists[] = { commandList.Get() };
+            commandQueue->ExecuteCommandLists(_countof(commandLists), commandLists);
+            ASSERT_SUCCESS(commandList->Reset(commandAllocators[frameIndex].Get(), nullptr));
+        }
+
+        void PresentFrame(uint32_t frameIndex) {
+            D3D12_RESOURCE_BARRIER resourceBarrier;
```

Comment:
I'm not sure what the semantics of D3D12 are, so this might not work. The code could be simplified by having SwapBuffers enqueue the following operations:

 - Transition from RENDER_TARGET to PRESENT
 - Present
 - Transition from PRESENT to RENDER_TARGET
 - Signal the fence.

Then SwapBuffers can be slightly simplified too to do the following:

 - Wait on the N -2 frame
 - Reset the allocators etc.
 - Enqueue work